File: 2016/day018c.py
from gmpy2 import popcount

inp = '.^^..^...^..^^.^^^.^^^.^^^^...^^^^.^.^.^^^.^^.^^.'
mask = (1 << len(inp)) - 1
row = int(inp.replace('.', '0').replace('^', '1'), 2)

visited = {}
safe = 100 - popcount(row)
i = 0
while True:
  if row in visited:
    print(i)
    die()
  row = ((row << 1) ^ (row >> 1)) & mask
  safe += 100 - popcount(row)
  i += 1
print(safe)

File: 2016/day018c.rs
use std::collections::HashMap;

const START: u64 = 0b0110010001001101110111011110001111010101110110110;
fn main() {
    let mut row = START;
    let mask: u64 = (1<<49) - 1;
    let mut visited: HashMap<u64, u64> = HashMap::new();
    let mut safe = row.count_zeros();
    for i in 0..u64::max_value() {
        println!("{:049b}", row);
        if visited.contains_key(&row) {
            println!("loop! {}", i);
            break;
        }
        visited.insert(row, i);
        row = ((row << 1) ^ (row >> 1)) & mask;
        safe += row.count_zeros();
    }
}
